aTalk/jni/openssl/android/armeabi-v7a/share/man/man3/HMAC.3
.\" Automatically generated by Pod::Man 4.11 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "HMAC 3"
.TH HMAC 3 "2021-03-25" "1.1.1k" "OpenSSL"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
HMAC, HMAC_CTX_new, HMAC_CTX_reset, HMAC_CTX_free, HMAC_Init, HMAC_Init_ex, HMAC_Update, HMAC_Final, HMAC_CTX_copy, HMAC_CTX_set_flags, HMAC_CTX_get_md, HMAC_size \&\- HMAC message authentication code
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\& #include <openssl/hmac.h>
\&
\& unsigned char *HMAC(const EVP_MD *evp_md, const void *key,
\&                     int key_len, const unsigned char *d, size_t n,
\&                     unsigned char *md, unsigned int *md_len);
\&
\& HMAC_CTX *HMAC_CTX_new(void);
\& int HMAC_CTX_reset(HMAC_CTX *ctx);
\&
\& int HMAC_Init_ex(HMAC_CTX *ctx, const void *key, int key_len,
\&                  const EVP_MD *md, ENGINE *impl);
\& int HMAC_Update(HMAC_CTX *ctx, const unsigned char *data, size_t len);
\& int HMAC_Final(HMAC_CTX *ctx, unsigned char *md, unsigned int *len);
\&
\& void HMAC_CTX_free(HMAC_CTX *ctx);
\&
\& int HMAC_CTX_copy(HMAC_CTX *dctx, HMAC_CTX *sctx);
\& void HMAC_CTX_set_flags(HMAC_CTX *ctx, unsigned long flags);
\& const EVP_MD *HMAC_CTX_get_md(const HMAC_CTX *ctx);
\&
\& size_t HMAC_size(const HMAC_CTX *e);
.Ve
.PP
Deprecated:
.PP
.Vb 4
\& #if OPENSSL_API_COMPAT < 0x10100000L
\& int HMAC_Init(HMAC_CTX *ctx, const void *key, int key_len,
\&               const EVP_MD *md);
\& #endif
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\s-1HMAC\s0 is a \s-1MAC\s0 (message authentication code), i.e. a keyed hash
function used for message authentication, which is based on a hash
function.
.PP
\&\s-1\fBHMAC\s0()\fR computes the message authentication code of the \fBn\fR bytes at
\&\fBd\fR using the hash function \fBevp_md\fR and the key \fBkey\fR which is
\&\fBkey_len\fR bytes long.
.PP
It places the result in \fBmd\fR (which must have space for the output of
the hash function, which is no more than \fB\s-1EVP_MAX_MD_SIZE\s0\fR bytes).
If \fBmd\fR is \s-1NULL,\s0 the digest is placed in a static array.  The size of
the output is placed in \fBmd_len\fR, unless it is \fB\s-1NULL\s0\fR. Note: passing a \s-1NULL\s0
value for \fBmd\fR  to use the static array is not thread safe.
.PP
\&\fBevp_md\fR is a message digest such as \fBEVP_sha1()\fR, \fBEVP_ripemd160()\fR etc. \s-1HMAC\s0 does
not support variable output length digests such as \fBEVP_shake128()\fR and
\&\fBEVP_shake256()\fR.
.PP
\&\fBHMAC_CTX_new()\fR creates a new \s-1HMAC_CTX\s0 in heap memory.
.PP
\&\fBHMAC_CTX_reset()\fR zeros an existing \fB\s-1HMAC_CTX\s0\fR and associated
resources, making it suitable for new computations as if it was newly
created with \fBHMAC_CTX_new()\fR.
.PP
\&\fBHMAC_CTX_free()\fR erases the key and other data from the \fB\s-1HMAC_CTX\s0\fR,
releases any associated resources and finally frees the \fB\s-1HMAC_CTX\s0\fR
itself.
.PP
The following functions may be used if the message is not completely
stored in memory:
.PP
\&\fBHMAC_Init_ex()\fR initializes or reuses a \fB\s-1HMAC_CTX\s0\fR structure to use the hash
function \fBevp_md\fR and key \fBkey\fR. If both are \s-1NULL,\s0 or if \fBkey\fR is \s-1NULL\s0
and \fBevp_md\fR is the same as the previous call, then the
existing key is
reused. \fBctx\fR must have been created with \fBHMAC_CTX_new()\fR before the first use
of an \fB\s-1HMAC_CTX\s0\fR in this function.
.PP
If \fBHMAC_Init_ex()\fR is called with \fBkey\fR \s-1NULL\s0 and \fBevp_md\fR is not the
same as the previous digest used by \fBctx\fR then an error is returned
because reuse of an existing key with a different digest is not supported.
.PP
\&\fBHMAC_Init()\fR initializes a \fB\s-1HMAC_CTX\s0\fR structure to use the hash
function \fBevp_md\fR and the key \fBkey\fR which is \fBkey_len\fR bytes
long.
.PP
\&\fBHMAC_Update()\fR can be called repeatedly with chunks of the message to
be authenticated (\fBlen\fR bytes at \fBdata\fR).
.PP
\&\fBHMAC_Final()\fR places the message authentication code in \fBmd\fR, which
must have space for the hash function output.
.PP
\&\fBHMAC_CTX_copy()\fR copies all of the internal state from \fBsctx\fR into \fBdctx\fR.
.PP
\&\fBHMAC_CTX_set_flags()\fR applies the specified flags to the internal EVP_MD_CTXs.
These flags have the same meaning as for \fBEVP_MD_CTX_set_flags\fR\|(3).
.PP
\&\fBHMAC_CTX_get_md()\fR returns the \s-1EVP_MD\s0 that has previously been set for the
supplied \s-1HMAC_CTX.\s0
.PP
\&\fBHMAC_size()\fR returns the length in bytes of the underlying hash function output.
.SH "RETURN VALUES"
.IX Header "RETURN VALUES"
\&\s-1\fBHMAC\s0()\fR returns a pointer to the message authentication code or \s-1NULL\s0 if
an error occurred.
.PP
\&\fBHMAC_CTX_new()\fR returns a pointer to a new \fB\s-1HMAC_CTX\s0\fR on success or
\&\fB\s-1NULL\s0\fR if an error occurred.
.PP
\&\fBHMAC_CTX_reset()\fR, \fBHMAC_Init_ex()\fR, \fBHMAC_Update()\fR, \fBHMAC_Final()\fR and
\&\fBHMAC_CTX_copy()\fR return 1 for success or 0 if an error occurred.
.PP
\&\fBHMAC_CTX_get_md()\fR return the \s-1EVP_MD\s0 previously set for the supplied \s-1HMAC_CTX\s0 or
\&\s-1NULL\s0 if no \s-1EVP_MD\s0 has been set.
.PP
\&\fBHMAC_size()\fR returns the length in bytes of the underlying hash function output
or zero on error.
.SH "CONFORMING TO"
.IX Header "CONFORMING TO"
\&\s-1RFC 2104\s0
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\s-1\fBSHA1\s0\fR\|(3), \fBevp\fR\|(7)
.SH "HISTORY"
.IX Header "HISTORY"
\&\fBHMAC_CTX_init()\fR was replaced with \fBHMAC_CTX_reset()\fR in OpenSSL 1.1.0.
.PP
\&\fBHMAC_CTX_cleanup()\fR existed in OpenSSL before version 1.1.0.
.PP
\&\fBHMAC_CTX_new()\fR, \fBHMAC_CTX_free()\fR and \fBHMAC_CTX_get_md()\fR are new in OpenSSL 1.1.0.
.PP
\&\fBHMAC_Init_ex()\fR, \fBHMAC_Update()\fR and \fBHMAC_Final()\fR did not return values in
OpenSSL before version 1.0.0.
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright 2000\-2020 The OpenSSL Project Authors. All Rights Reserved.
.PP
Licensed under the OpenSSL license (the \*(L"License\*(R").  You may not use
this file except in compliance with the License.  You can obtain a copy
in the file \s-1LICENSE\s0 in the source distribution or at
<https://www.openssl.org/source/license.html>.


aTalk/ReleaseNotes.txt
Project aTalk-Android Release Notes
=========================================================================
Author:         cmeng
Upload Date:	05/16/2021
Version:		2.6.1 (206010)
- AndroidCodec: Change implementation to remove all deprecated MediaCodec methods
  codecInputs = codec.getInputBuffers();
  codecOutputs = codec.getOutputBuffers();
- Upgrade libvpx to v 1.10.0; and update VPX.java static constant per libvpx v1.10.0
- Clean up VPX constants, VP8 codec source; VP9 codec clean up but still not working
- Upgrade ffmpeg-x264 build to use x264-161.3049 i.e. git clone https://code.videolan.org/videolan/x264.git --branch stable
- Tested working for ffmpeg v3.4.8 without source modification
- Upgraded ffmpeg version v4.4; need the following changes
  * Update all deprecated methods in FFmpeg.c
  * need to clean up Ffmpeg.java constants as per ffmpeg v4.4 release, e.g. CODEC_ID_VP8 = 139 else
    net.sf.fmj.media.Log.error()  Unable to handle format: H264/RTP, fmtps={profile-level-id=42E01f,packetization-mode=1}
  * Need to include FFmpeg.avframe_set_properties(avFrame, FFmpeg.PIX_FMT_YUV420P, mWidth, mHeight);
- Upgraded openssl to v1.1.1k and uses prebuilt ndk
- Remove unnecessary calls to deprecated functions in FFmpeg.java
  av_register_all(); avcodec_register_all(); avfilter_register_all();
- Update all jni/static_library_build scripts for ffmpeg-x264 and libvpx per latest deployed versions;
  directly install the generated static libraries to aTalk jni directory for build
- Enhance x264 encoder to support android device rotation
- cmeng (20210504): Call with/initiated by Conversation may get terminated abruptly, when h264 video codec is used
  CallPeerMediaHandel#getDefaultDeivce: display toast message and let it ends with NPE for simple fix;
  otherwise more unrelated error messages will display
- Upgrade ExoPlayer to version 2.13.3
- MediaExoPLayer activity is set to android:configChanges="keyboardHidden|orientation|screenSize",
  and orientation change is now handled by exoPlayer with smooth audio/video playback on screen rotation.
- JCenter deprecation and end of service: https://developer.android.com/studio/build/jcenter-migration
  https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/
  Replace jCenter with mavenCentral amd omit jCenter from gradle

=========================================================================
Author:         cmeng
Upload Date:	04/05/2021
Version:		2.6.0 (206000)
- Check for user valid ipAddress entry in STUN/TURN setting configuration
- Cancel vibrator immediately upon call accepted by user i.e. JingleMessageHelper.sendJingleAccept()
  and showToaste message "Connecting caller" on received JingleProceed message
- Remove ProgressDialogFragment.showProgressDialog() in AndroidCallUtil#createCall(). Does not have chance to display and
  often have error in DialogActivity.waitForDialogOpened(dialogId), leaving an un-closeable (by user) dialog fragment
- Android throws the following error when camera resolution is set to 720x480 in portrait mode;
  522240 is for landscape size calculated by by PreviewStream#calcYV12Size:
  E/Camera-JNI: Callback buffer was too small! Expected 522240 bytes, but got 518400 bytes!
  Reason: aTalk always set to use landscape (0 deg) in camera preview; see PreviewStream#calcYV12Size()
- Default enable VP8 codec and set as first priority. Default video resolution to "720x480"
- RemoteVideoLayout#setVideoPreferredSize() Size change algorithm must use preferredSize and videoSize ratio compare;
  Otherwise, not null remoteVideoView will also return false for sizeChange when remote video dimension changes
  due to device orientation change; removed this test as it is unnecessary.

======== Video call with device orientation change ==========
Three approaches: #3 is selected for final implementation to give the best result:
1: In VideoCallActivity#onCreate, use setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_XXX)
   When sender changes orientation, recipient needs to change his device orientation to make the remote video view upright -> then sender remote view becomes OK.
   The video stream encryption with Replay protect should work properly as there is not interruption in stream sending.
   However when switch to back camera, the remote video is inverted. Both sender & recipient must switch to back cameras to work properly.
2. Approach aborted due to complexity and NewReceiveStreamEvent is not always gets triggered - reliability issue
   Do not send content-modify when the sender device orientation change as there is no change in senders attribute.
   see https://xmpp.org/extensions/xep-0166.html#def-action-content-modify
   The content-modify action is used to change the direction of an existing content definition through modification of the 'senders' attribute.
3. Use VideoActivity#onConfigurationChanged() event, and let app handles all the required changes i.e.
   a. Local content view redraw
   b. Need to check camera rotation; front/back to correctly transform video data for streaming; to avoid stretch and orientation
   b. Rotated camera preview sending, requires recipient CODEC to auto detect video stream size change. Currently only VP8 can support

======== Video call with camera switching ==========
- see PreviewStream#switchCamera(); release current camera and start user selected camera.
  # MediaStreamImpl#setDevice: remove check for device change in "if ((deviceSession == null) || (deviceSession.getDevice() != device))"
  # Seems not necessary after the above change, so reinsert back:
    Block DeviceConfiguration#setVideoCaptureDevice to firePropertyChange(VIDEO_CAPTURE_DEVICE, oldDevice, device);
    to avoid triggering of JIngle messages (content modify) sending

- The above two new implementations should improve SRTP replay protection reliability on screen rotation or switching camera


aTalk/jni/openssl/Hmac.c
/*
 * Copyright @ 2015 - present 8x8, Inc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <stdint.h>
#include <stdlib.h>
#include <openssl/evp.h>
#include <openssl/hmac.h>
#include "Hmac.h"

/*
 * Class:     org_atalk_impl_neomedia_transform_srtp_crypto_OpenSslHmac
 * Method:    EVP_MD_size
 * Signature: (J)I
 */
JNIEXPORT jint JNICALL
Java_org_atalk_impl_neomedia_transform_srtp_crypto_OpenSslHmac_EVP_1MD_1size
  (JNIEnv *env, jclass clazz, jlong md)
{
    return EVP_MD_size((const EVP_MD *)(intptr_t) md);
}

/*
 * Class:     org_atalk_impl_neomedia_transform_srtp_crypto_OpenSslHmac
 * Method:    EVP_sha1
 * Signature: ()J
 */
JNIEXPORT jlong JNICALL
Java_org_atalk_impl_neomedia_transform_srtp_crypto_OpenSslHmac_EVP_1sha1
  (JNIEnv *env, jclass clazz)
{
    return (jlong)(intptr_t) EVP_sha1();
}

/*
 * Class:     org_atalk_impl_neomedia_transform_srtp_crypto_OpenSslHmac
 * Method:    HMAC_CTX_create
 * Signature: ()J
 */
JNIEXPORT jlong JNICALL
Java_org_atalk_impl_neomedia_transform_srtp_crypto_OpenSslHmac_HMAC_1CTX_1create
  (JNIEnv *env, jclass clazz)
{
/* OpenSSL 1.1.0 made HMAC_CTX an opaque structure, which must be allocated
   using HMAC_CTX_new.  But this function doesn't exist in OpenSSL 1.0.x. */
#if OPENSSL_VERSION_NUMBER < 0x10100000L || LIBRESSL_VERSION_NUMBER
    HMAC_CTX *ctx = malloc(sizeof(HMAC_CTX));

    if (ctx)
        HMAC_CTX_init(ctx);

#else
    HMAC_CTX *ctx = HMAC_CTX_new();

#endif

    return (jlong) (intptr_t) ctx;
}

/*
 * Class:     org_atalk_impl_neomedia_transform_srtp_crypto_OpenSslHmac
 * Method:    HMAC_CTX_destroy
 * Signature: (J)V
 */
JNIEXPORT void JNICALL
Java_org_atalk_impl_neomedia_transform_srtp_crypto_OpenSslHmac_HMAC_1CTX_1destroy
  (JNIEnv *env, jclass clazz, jlong ctx)
{
    HMAC_CTX *ctx_ = (HMAC_CTX *) (intptr_t) ctx;

#if OPENSSL_VERSION_NUMBER < 0x10100000L || LIBRESSL_VERSION_NUMBER
    HMAC_CTX_cleanup(ctx_);
    free(ctx_);

#else
    HMAC_CTX_free(ctx_);

#endif
}

/*
 * Class:     org_atalk_impl_neomedia_transform_srtp_crypto_OpenSslHmac
 * Method:    HMAC_Final
 * Signature: (J[BII)I
 */
JNIEXPORT jint JNICALL
Java_org_atalk_impl_neomedia_transform_srtp_crypto_OpenSslHmac_HMAC_1Final
  (JNIEnv *env, jclass clazz, jlong ctx, jbyteArray md, jint mdOff, jint mdLen)
{
    jbyte *md_ = (*env)->GetPrimitiveArrayCritical(env, md, NULL);
    int i;

    if (md_)
    {
        unsigned int len = mdLen;

        i
            = HMAC_Final(
                    (HMAC_CTX *) (intptr_t) ctx,
                    (unsigned char *) (md_ + mdOff),
                    &len);
        (*env)->ReleasePrimitiveArrayCritical(env, md, md_, 0);
        i = i ? len : -1;
    }
    else
    {
        i = -1;
    }
    return i;
}

/*
 * Class:     org_atalk_impl_neomedia_transform_srtp_crypto_OpenSslHmac
 * Method:    HMAC_Init_ex
 * Signature: (J[BIJJ)Z
 */
JNIEXPORT jboolean JNICALL
Java_org_atalk_impl_neomedia_transform_srtp_crypto_OpenSslHmac_HMAC_1Init_1ex
  (JNIEnv *env, jclass clazz, jlong ctx, jbyteArray key, jint keyLen, jlong md, jlong impl)
{
    jbyte *key_;
    jboolean ok;

    if (key)
    {
        key_ = (*env)->GetPrimitiveArrayCritical(env, key, NULL);
        ok = key_ ? JNI_TRUE : JNI_FALSE;
    }
    else
    {
        key_ = NULL;
        ok = JNI_TRUE;
    }
    if (JNI_TRUE == ok)
    {
	    ok = HMAC_Init_ex(
	        (HMAC_CTX *)(intptr_t) ctx,
	        (const void *)  (intptr_t) key_,
	        keyLen,
	        (const EVP_MD *)(intptr_t) md,
	        (ENGINE *)(intptr_t) impl);
	    if (key_)
	    	(*env)->ReleasePrimitiveArrayCritical(env, key, key_, JNI_ABORT);
    }
    return ok;
}

/*
 * Class:     org_atalk_impl_neomedia_transform_srtp_crypto_OpenSslHmac
 * Method:    HMAC_Update
 * Signature: (J[BII)Z
 */
JNIEXPORT jboolean JNICALL
Java_org_atalk_impl_neomedia_transform_srtp_crypto_OpenSslHmac_HMAC_1Update
  (JNIEnv *env, jclass clazz, jlong ctx, jbyteArray data, jint offset, jint len)
{
    jbyte *data_ = (*env)->GetPrimitiveArrayCritical(env, data, NULL);
    jboolean ok;

    if (data_)
    {
	    ok = HMAC_Update(
	        (HMAC_CTX *)(intptr_t) ctx,
	        (const unsigned char *) (data_ + offset),
	        len);
	    (*env)->ReleasePrimitiveArrayCritical(env, data, data_, JNI_ABORT);
  	}
    else
    {
        ok = JNI_FALSE;
    }
    return ok;
}
